### PR TITLE
Node 1234: api pagination transactions address

### PR DESCRIPTION
**Pull request by dvshur**

### CODE REVIEW

File: src/test/scala/com/wavesplatform/state/RollbackSpec.scala [hunk 1]
```diff
@@ -92,8 +92,8 @@ class RollbackSpec extends FreeSpec with Matchers with WithState with Transactio
               ))
           }
 
-          val stransactions1 = d.addressTransactions(sender).sortBy(_._2.timestamp)
-          val rtransactions1 = d.addressTransactions(recipient).sortBy(_._2.timestamp)
+          val stransactions1 = d.addressTransactions(sender).getOrElse(Seq.empty).sortBy(_._2.timestamp)
```
Comment (ismagin) [inline src/test/scala/com/wavesplatform/state/RollbackSpec.scala]: not sure getOrElse is good here, what if both asserted version and expected result fail? the test would pass.
maybe use `.explicitGet()` here?

File: src/test/scala/com/wavesplatform/database/LevelDBWriterSpec.scala [hunk 2]
```diff
@@ -154,6 +155,54 @@ class LevelDBWriterSpec extends FreeSpec with Matchers with WithDB with RequestG
...
     }
 
+    "return Left if fromId argument is a non-existent transaction" in {
+      baseTest(time => preconditions(time.correctedTime())) { (writer, account) =>
+        val nonExistentTxId = GenesisTransaction.create(account, ENOUGH_AMT, 1).explicitGet().id()
+
+        val txs = writer
+          .addressTransactions(account.toAddress, Set(TransferTransactionV1.typeId), 3, Some(nonExistentTxId))
+
+        txs shouldBe Left(s"Transaction $nonExistentTxId does not exist")
```
Comment (ismagin) [inline src/test/scala/com/wavesplatform/database/LevelDBWriterSpec.scala]: `txs should produce("does not exist")` is another way to assert `Left`, but not in a strict way. Doesn't matter in this case

File: src/test/scala/com/wavesplatform/database/LevelDBWriterSpec.scala [hunk 3]
```diff
@@ -154,6 +155,54 @@ class LevelDBWriterSpec extends FreeSpec with Matchers with WithDB with RequestG
...
+
+    "return an empty Seq when paginating from the last transaction" in {
+      baseTest(time => preconditions(time.correctedTime())) { (writer, account) =>
+        val txs = writer
+          .addressTransactions(account.toAddress, Set(TransferTransactionV1.typeId), 2, None)
+          .getOrElse(Seq.empty)
+
+        val txsFromLast = writer
+          .addressTransactions(account.toAddress, Set(TransferTransactionV1.typeId), 2, Some(txs.last._2.id()))
+          .getOrElse(txs) // should never return this value
```
Comment (ismagin) [inline src/test/scala/com/wavesplatform/database/LevelDBWriterSpec.scala]: if this should not happen but fail instead, `.explicitGet()` may be more suitable

File: src/test/scala/com/wavesplatform/state/RollbackSpec.scala (hunk 1 above)
Comment (dvshur) [reply src/test/scala/com/wavesplatform/state/RollbackSpec.scala]: +

File: src/test/scala/com/wavesplatform/database/LevelDBWriterSpec.scala (hunk 3 above)
Comment (dvshur) [reply src/test/scala/com/wavesplatform/database/LevelDBWriterSpec.scala]: +

File: src/main/scala/com/wavesplatform/api/http/leasing/LeaseApiRoute.scala [hunk 4]
```diff
@@ -73,7 +73,8 @@ case class LeaseApiRoute(settings: RestAPISettings, wallet: Wallet, blockchain:
         case Left(e) => ApiError.fromValidationError(e)
         case Right(a) =>
           blockchain
-            .addressTransactions(a, Set(LeaseTransactionV1.typeId), Int.MaxValue, 0)
+            .addressTransactions(a, Set(LeaseTransactionV1.typeId), Int.MaxValue, None)
+            .getOrElse(Seq.empty)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/api/http/leasing/LeaseApiRoute.scala]: explicitGet()

File: src/main/scala/com/wavesplatform/state/package.scala [hunk 5]
```diff
@@ -77,12 +118,14 @@ package object state {
 
     def aliasesOfAddress(address: Address): Seq[Alias] =
       blockchain
-        .addressTransactions(address, Set(CreateAliasTransactionV1.typeId), Int.MaxValue, 0)
+        .addressTransactions(address, Set(CreateAliasTransactionV1.typeId), Int.MaxValue, None)
+        .getOrElse(Seq.empty)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/package.scala]: explicitGet

File: src/main/scala/com/wavesplatform/state/package.scala [hunk 6]
```diff
@@ -77,12 +118,14 @@ package object state {
...
         .collect { case (_, a: CreateAliasTransaction) => a.alias }
 
     def activeLeases(address: Address): Seq[(Int, LeaseTransaction)] =
       blockchain
-        .addressTransactions(address, Set(LeaseTransactionV1.typeId), Int.MaxValue, 0)
+        .addressTransactions(address, Set(LeaseTransactionV1.typeId), Int.MaxValue, None)
+        .getOrElse(Seq.empty)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/package.scala]: explicitGet

File: src/main/scala/com/wavesplatform/state/patch/CancelLeaseOverflow.scala [hunk 7]
```diff
@@ -14,9 +14,10 @@ object CancelLeaseOverflow extends ScorexLogging {
     addressesWithLeaseOverflow.keys.foreach(addr => log.info(s"Resetting lease overflow for $addr"))
 
     val leasesToCancel = addressesWithLeaseOverflow.keys.flatMap { a =>
-      blockchain.addressTransactions(a, Set(LeaseTransactionV1.typeId), Int.MaxValue, 0).collect {
-        case (_, lt: LeaseTransaction) if lt.sender.toAddress == a => lt.id()
-      }
+      blockchain
+        .addressTransactions(a, Set(LeaseTransactionV1.typeId), Int.MaxValue, None)
+        .getOrElse(Seq.empty)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state/patch/CancelLeaseOverflow.scala]: explicitGet

File: src/main/scala/com/wavesplatform/api/http/TransactionsApiRoute.scala [hunk 8]
```diff
@@ -45,42 +44,51 @@ case class TransactionsApiRoute(settings: RestAPISettings,
 
   override lazy val route =
     pathPrefix("transactions") {
-      unconfirmed ~ addressLimit ~ info ~ sign ~ calculateFee ~ broadcast
+      unconfirmed ~ addressLimit ~ info ~ sign ~ calculateFee ~ broadcast ~ complete(StatusCodes.NotFound)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/api/http/TransactionsApiRoute.scala]: let's remove this

**Comment by ismagin** [comment]
merged in another PR

**Pull request closed without merging**